ledger/smt.go

```diff
@@ -0,0 +1,402 @@
+		batch), nil}
+}
+
+// splitKeys devides the array of keys into 2 so they can update left and right branches in parallel
+func (s *SMT) splitKeys(keys [][]byte, height int) ([][]byte, [][]byte) {
+	for i, key := range keys {
+		if bitIsSet(key, height) {
+			return keys[:i], keys[i:]
+		}
+	}
+	return keys, nil
+}
+
+// maybeAddShortcutToKV adds a shortcut key to the keys array to be updated.
+// this is used when a subtree containing a shortcut node is being updated
+func (s *SMT) maybeAddShortcutToKV(keys, values [][]byte, shortcutKey, shortcutVal []byte) ([][]byte, [][]byte) {
+	newKeys := make([][]byte, 0, len(keys)+1)
+	newVals := make([][]byte, 0, len(keys)+1)
+
+	if bytes.Compare(shortcutKey, keys[0]) < 0 {
+		newKeys = append(newKeys, shortcutKey)
+		newKeys = append(newKeys, keys...)
+		newVals = append(newVals, shortcutVal)
+		newVals = append(newVals, values...)
+	} else if bytes.Compare(shortcutKey, keys[len(keys)-1]) > 0 {
+		newKeys = append(newKeys, keys...)
+		newKeys = append(newKeys, shortcutKey)
+		newVals = append(newVals, values...)
+		newVals = append(newVals, shortcutVal)
+	} else {
+		higher := false
+		for i, key := range keys {
+			if bytes.Equal(shortcutKey, key) {
+				// the shortcut keys is being updated
+				return keys, values
+			}
+			if !higher && bytes.Compare(shortcutKey, key) > 0 {
+				higher = true
+				continue
+			}
+			if higher && bytes.Compare(shortcutKey, key) < 0 {
+				// insert shortcut in slices
+				newKeys = append(newKeys, keys[:i]...)
+				newKeys = append(newKeys, shortcutKey)
+				newKeys = append(newKeys, keys[i:]...)
+				newVals = append(newVals, values[:i]...)
+				newVals = append(newVals, shortcutVal)
+				newVals = append(newVals, values[i:]...)
+				break
+			}
+		}
+	}
+	return newKeys, newVals
+}
+
+// loadChildren looks for the children of a node.
+// if the node is not stored in cache, it will be loaded from db.
+func (s *SMT) loadChildren(root []byte, height, iBatch int, batch [][]byte) ([][]byte, int, []byte, []byte, bool,
+	error) {
+	isShortcut := false
+	if height%4 == 0 {
+		if len(root) == 0 {
+			// create a new default batch
+			batch = make([][]byte, 31)
+			batch[0] = []byte{0}
+		} else {
+			var err error
+			batch, err = s.loadBatch(root[:HashLength])
+			if err != nil {
+				return nil, 0, nil, nil, false, err
+			}
+		}
+		iBatch = 0
+		if batch[0][0] == 1 {
+			isShortcut = true
+		}
+	} else if len(batch[iBatch]) != 0 && batch[iBatch][HashLength] == 1 {
+		isShortcut = true
+	}
+	return batch, iBatch, batch[2*iBatch+1], batch[2*iBatch+2], isShortcut, nil
+}
+
+// loadBatch fetches a batch of nodes in cache or db
+func (s *SMT) loadBatch(root []byte) ([][]byte, error) {
+	var node Hash
+	copy(node[:], root)
+
+	// checking updated nodes is useful if get() or update() is called twice in a row without db commit
+	s.db.updatedMux.RLock()
+	val, exists := s.db.updatedNodes.Get(node)
+	s.db.updatedMux.RUnlock()
+	if exists {
+		if s.atomicUpdate {
```

Comment:
s.atomicUpdate is accessed without a lock, is that right?

Reply:
loadBatch is unexported, and any exported function which calls loadBatch directly or indirectly uses the lock.

Reply:
Yes, that mutex is specifically for the cache.  I can add a mutex for the AtomicUpdate bool if that would help...

Reply:
The right fix for the atomicUpdate field then is to probably use atomic read/wrties on it for all accesses rather than a mutex. That should keep the race detection happy.